src/Illuminate/Session/Middleware/StartSession.php

```diff
@@ -47,45 +40,23 @@ public function __construct(SessionManager $manager)
      */
     public function handle($request, Closure $next)
     {
-        $this->sessionHandled = true;
-
-        // If a session driver has been configured, we will need to start the session here
-        // so that the data is ready for an application. Note that the Laravel sessions
-        // do not make use of PHP "native" sessions in any way since they are crappy.
         if ($this->sessionConfigured()) {
             $request->setLaravelSession(
                 $session = $this->startSession($request)
             );
 
             $this->collectGarbage($session);
-        }
 
-        $response = $next($request);
-
-        // Again, if the session has been configured we will need to close out the session
-        // so that the attributes may be persisted to some storage medium. We will also
-        // add the session identifier cookie to the application response headers now.
-        if ($this->sessionConfigured()) {
             $this->storeCurrentUrl($request, $session);
 
-            $this->addCookieToResponse($response, $session);
-        }
+            $response = $next($request);
 
-        return $response;
-    }
+            $this->addCookieToResponse($response, $session);
 
-    /**
-     * Perform any final actions for the request lifecycle.
-     *
-     * @param  \Illuminate\Http\Request  $request
-     * @param  \Symfony\Component\HttpFoundation\Response  $response
-     * @return void
-     */
-    public function terminate($request, $response)
-    {
-        if ($this->sessionHandled && $this->sessionConfigured() && ! $this->usingCookieSessions()) {
             $this->manager->driver()->save();
         }
+
+        return $response ?? $next($request);
```

Comment:
Wouldn't this execute the action twice in case it returned a false-y (null, false, empty string, etc) above?

Reply:
I'm wrong, $response cannot be false-y here since we've called addCookieToResponse that requires it to be a Response.